substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java

```diff
@@ -1510,7 +1529,20 @@ protected int buildImage(List<String> javaArgs, LinkedHashSet<Path> cp, LinkedHa
         Function<Path, Path> substituteModulePath = useBundle() ? bundleSupport::substituteModulePath : Function.identity();
         List<Path> substitutedImageModulePath = imagemp.stream().map(substituteModulePath).toList();
 
-        Map<String, Path> modules = listModulesFromPath(javaExecutable, Stream.concat(mp.stream(), imagemp.stream()).distinct().toList());
+        List<String> mainClassArg = config.getGeneratorMainClass();
+        /*
+         * We are using the image generator for --list-modules. That cannot handle the same modules
+         * on the module path (--module-path) and image module path (-imagemp). Only pass modules
+         * via the image module path if it isn't already on the module path. Note that the path to
+         * the individual jars might be different between module path and image module path, so we
+         * use the file name for filtering.
+         */
+        Set<Path> imagempForListMods = imagemp.stream().filter(Predicate.not(a -> {
```

Comment:
Instead of this hack we should find out why in the case of running those sulong tests we do get a module on the `-imagemp` that we already have on the module-path of the builder. I would expect this to be solved on mx level.

Reply:
Fair enough, who can I get involved to look at this from the sulong side? My main aim is to get the Mandrel build unbroken, which is the case for about a month now and this fix is dragging along longer than I'd have thought. :-(

Here is what I know...

For the sulong tests you get this modulepath:
```
--module-path \
/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/truffle/truffle-api.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/truffle/truffle-runtime.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/truffle/truffle-compiler.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/truffle/builder/truffle-runtime-svm.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/svm/builder/native-image-base.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/svm/builder/svm.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/svm/builder/svm-llvm.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/svm/builder/javacpp-shadowed.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/svm/builder/llvm-wrapper-shadowed.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/svm/builder/objectfile.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/svm/builder/llvm-platform-specific-shadowed.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/svm/builder/javacpp-platform-specific-shadowed.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/svm/builder/pointsto.jar
```
and this image module path:

```
-imagemp \
/home/runner/work/graal/graal/truffle/mxbuild/jdk21/dists/jdk21/truffle-api.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/truffle/builder/truffle-runtime-svm.jar:/home/runner/work/graal/graal/sdk/mxbuild/jdk21/dists/jdk17/polyglot.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/svm/macros/junit/junit-support.jar:/home/runner/work/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_F06648F35C_JAVA21/graalvm-f06648f35c-java21-24.0.0-dev/lib/svm/library-support.jar
```

So you get `truffle-api.jar` and `truffle-runtime-svm.jar` as duplicates causing the issue. I don't know why :-/

I'm sort of out of my depth when it comes to `mx`, let alone `sulong`. So again, who would be able to help? Or better, yet, could we get this fix in, *as is*, first and file a follow-up issue somewhere to do this cleanup? We really want to get the Mandrel build working again and this patch is the missing piece. Thoughts?